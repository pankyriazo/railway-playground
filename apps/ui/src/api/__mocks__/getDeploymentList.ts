import { GetDeploymentListResponse } from "@/api/types";

export const mockGetDeploymentList = {
  data: {
    deployments: {
      edges: [
        {
          node: {
            id: "4b049de0-1ce0-4e81-8036-26b252e3285c",
            status: "SUCCESS",
            updatedAt: "2024-03-17T08:50:14.973Z",
            environmentId: "632da482-c5b1-4570-87af-9ea225c335cb",
          },
        },
        {
          node: {
            id: "38b68be4-f86d-4987-8c48-dee293fda404",
            status: "SUCCESS",
            updatedAt: "2024-03-16T19:41:25.747Z",
            environmentId: "1beb4373-6b1f-43cb-8150-0e7214acea82",
          },
        },
      ],
    },
  },
} as const satisfies GetDeploymentListResponse;

export const mockGetDeploymentListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetDeploymentListResponse;
