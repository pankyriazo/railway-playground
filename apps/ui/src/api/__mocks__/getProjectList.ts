import { GetProjectListResponse } from "@/api/types";

export const mockGetProjectList = {
  data: {
    me: {
      projects: {
        edges: [
          {
            node: {
              id: "a20a0d44-a6de-4804-acf1-58d436570fd7",
              name: "nauseating-cup",
              services: {
                edges: [
                  {
                    node: {
                      id: "0e62e7bf-2008-4452-bd32-e57cec3bb9a2",
                      name: "fascinated-motion",
                    },
                  },
                  {
                    node: {
                      id: "7ab42902-aae8-47b7-9976-81496dc61834",
                      name: "Postgres",
                    },
                  },
                ],
              },
              plugins: {
                edges: [],
              },
              environments: {
                edges: [
                  {
                    node: {
                      id: "1beb4373-6b1f-43cb-8150-0e7214acea82",
                      name: "production",
                    },
                  },
                  {
                    node: {
                      id: "632da482-c5b1-4570-87af-9ea225c335cb",
                      name: "staging",
                    },
                  },
                ],
              },
            },
          },
        ],
      },
    },
  },
} as const satisfies GetProjectListResponse;

export const mockGetProjectListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetProjectListResponse;
