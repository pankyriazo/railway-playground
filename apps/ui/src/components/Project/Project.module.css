.container {
  background-color: var(--color-bg);
  border: 1px solid var(--color-border--900);
  border-radius: var(--border-radius);
  padding: var(--spacing-md);
  transition: border-color var(--transition);
}
.container:hover {
  border-color: var(--color-border--500);
}

.title {
  color: var(--color-text--900);
  font-size: var(--font-size-lg);
  font-weight: var(--font-weight-bold);
  margin-bottom: var(--spacing-sm);
}

.services {
  margin-top: var(--spacing-lg);
}
