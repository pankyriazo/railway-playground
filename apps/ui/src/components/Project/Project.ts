import { Node, Project as ProjectType } from "@monorepo/types";
import { Component } from "@/types";
import Services from "@/components/Services";
import EnvironmentSelect from "@/components/EnvironmentSelect";

import styles from "./Project.module.css";

class Project extends Component<HTMLLIElement> {
  private _project: ProjectType;
  private _environment: Node | null = null;

  constructor(project: ProjectType) {
    super();

    this._element = document.createElement("li");
    this._element.id = `project-${project.id}`;
    this._project = project;
    this._environment = project.environments.edges[0]?.node ?? null;

    this.render();
  }

  private _onEnvironmentChange(environment: Node) {
    this._environment = environment;
    this.render();
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.dataset.testid = "project";
    this._element.className = styles.container!;

    const title = document.createElement("h3");
    title.dataset.testid = "project__title";
    title.className = styles.title!;
    title.textContent = this._project.name;
    this._element.appendChild(title);

    if (!this._environment) {
      return;
    }

    const environmentInstance = new EnvironmentSelect({
      options: this._project.environments,
      selected: this._environment,
      onChange: this._onEnvironmentChange.bind(this),
    });
    const environment = document.createElement("div");
    environment.className = styles.environment!;
    environment.appendChild(environmentInstance.element);
    this._element.appendChild(environment);

    const servicesInstance = new Services({
      project: this._project,
      services: this._project.services,
      environment: this._environment,
    });
    const services = document.createElement("div");
    services.className = styles.services!;
    services.appendChild(servicesInstance.element);
    this._element.appendChild(services);
  }
}

export default Project;
