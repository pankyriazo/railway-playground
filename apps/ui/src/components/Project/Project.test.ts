import { mockComponent, render } from "@/test-helpers";
import { mockGetProjectList } from "@/api/__mocks__/getProjectList";
import Project from "./Project";

jest.mock("@/components/Services", () => mockComponent("Services"));
jest.mock("@/components/EnvironmentSelect", () =>
  mockComponent("EnvironmentSelect")
);

describe("Project", () => {
  const mockProps = mockGetProjectList.data.me.projects.edges[0].node;

  beforeEach(() => {
    render(Project, mockProps);
  });

  it("should render", () => {
    const project = document.querySelector("[data-testid=project]");

    expect(project).toMatchSnapshot();
  });
});
