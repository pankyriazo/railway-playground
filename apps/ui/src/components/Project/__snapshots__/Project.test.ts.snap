// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project should render 1`] = `
<li
  class="container"
  data-testid="project"
  id="project-a20a0d44-a6de-4804-acf1-58d436570fd7"
>
  <h3
    class="title"
    data-testid="project__title"
  >
    nauseating-cup
  </h3>
  <div
    class="environment"
  >
    <div
      data-component="EnvironmentSelect"
    />
  </div>
  <div
    class="services"
  >
    <div
      data-component="Services"
    />
  </div>
</li>
`;
